### ISSUE TITLE
Accessing certain properties on Window.prototype throws under the Shadow DOM polyfill

### ISSUE THREAD

**Issue by rictic** [webcomponents/webcomponentsjs#365]
Added a test demonstrating this issue here: https://github.com/rictic/webcomponentsjs/commit/87c78e8c7840175f2a3f060032db05fdc8365273

This behavior is currently causing issues for Google Play Music on Microsoft Edge, but I can reproduce the error with Firefox as well.

**Comment by jmesserly** [comment]
it happens on all browsers as far as I can tell. Here's a jsbin.
https://jsbin.com/xucupegaki/edit?html,output

The issue is with `Window.prototype.localStorage`, which accesses a getter method without the correct `this`. I'm not sure this can be fixed, or even that the behavior is incorrect.

EcmaScript 5.1 introduced getters, in general it's not safe to access them from the prototype like that.

If you're trying to reflect on Window.prototype, the safe way would be to use Object.getOwnPropertyNames/Object.getOwnPropertyDescriptor.

Here's an example of the issue with normal DOM APIs:
```
$ HTMLElement.prototype.innerHTML
VM708:1 Uncaught TypeError: Illegal invocation(…)
```

It looks like most browsers don't put anything interesting on Window.prototype, instead they go straight on the `window` instance (e.g. `Object.getOwnPropertyNames(window)`)

The thing is ... we're *very* limited in what we can do on Window (and Document, for that matter). I'll poke around, but I'm not sure we'll be able to do anything here. This is the sort of thing that falls into the ShadowDOM polyfill's known limitations.

**Comment by jmesserly** [comment]
Related bug with some discussion:
https://bugzilla.mozilla.org/show_bug.cgi?id=943065

if you're sufficiently motivated, the way to fix this would be to change how src/wrappers/Window.js works, and avoid the Window.prototype, instead patching things on `window`. Not sure if it will work, but might be worth a shot. Where they landed with WebIDL was leaving things on window instance, whereas all other accessors are moving to the appropriate prototype*.

(code should be hardened against the `HTMLElement.prototype.innerHTML` example ... almost all properties will be working that way. Except on `window`)

**Comment by rictic** [comment]
So the other half of this issue is this code in closure library: https://github.com/google/closure-library/blob/master/closure/goog/async/nexttick.js#L88

I suppose we could add code there to catch the error when it tries to access Window.prototype.setImmediate.

**Comment by jmesserly** [comment]
hmmmm. that's interesting. I wouldn't expect setImmediate to trigger the bug... let me see what's up there...

**Comment by jmesserly** [comment]
hmm, not seeing the setImmediate issue (Firefox or Chrome or Safari with SD polyfill forced to on). Though I can definitely imagine Window.prototype is doing strange things :)
I'll see if I can get a hold of an IE Edge browser to test with

**Comment by jmesserly** [comment]
Is that the same error with accessing this.__impl*__ where `this` is undefined?

**Comment by rictic** [comment]
I'm having trouble reproducing the setImmediate bug in this test, but if I go to a site in Edge that includes webcomponents.js, like polymer-project.org, and I open up the console and type `Window.prototype.setImmediate` then I get the error `Unable to get property 'setImmediate' of undefined or null reference`, just like I get when I do `Window.prototype.alert`. I can't look at `Window.prototype` directly, as Edge's dev tools throw several hundred `Unable to get property '$propName' of undefined or null reference` errors when I try, but `'setImmediate' in Window.prototype` is `true`.

**Comment by azakus** [comment]
I only see the `setImmediate` issue when using v0.6.3. Using master or v0.7.x do not show this issue.
We should see about making another maintenance release for Polymer 0.5 that uses WC 0.7

**Comment by azakus** [comment]
I do still see the error for `localStorage`, but we can fix that up in the `Window` wrapper.

**Comment by jmesserly** [comment]
> I only see the setImmediate issue when using v0.6.3. Using master or v0.7.x do not show this issue.
We should see about making another maintenance release for Polymer 0.5 that uses WC 0.7

oh, nice find! yeah, I was testing against master

> I do still see the error for localStorage, but we can fix that up in the Window wrapper.

@azakus curious what you were thinking? Just patching localStorage, or all getters? With the registerWrapper call here:
https://github.com/webcomponents/webcomponentsjs/blob/8289c09323b2fd2aa9e827c1eb34db96d047977e/src/ShadowDOM/wrappers/Window.js#L90

IIRC, it will end up creating getters on Window.prototype from all props on `window`.

Maybe we could hook up special getters?
```
if (!(this instanceof Window)) return undefined;
... existing logic ...
```
Feels a bit odd, but that would let us pretend those props aren't there. Perf cost though.
If we just wanted to do it for localStorage specifically, that'd be easier.

More radical idea would be to change how `window` wrapping works, so the instance is the place with all of the properties, not the the (polyfilled) Window.prototype. I presume there was a reason Erik originally used `Window.prototype`. However `window` would be more correct based on my understanding of the current specs.

**Comment by azakus** [comment]
Hmm, I'm concerned that changing the `Window.prototype` accessors will break something, but I don't what I would be looking for.

Maybe the minimal `instanceof` patching is enough for now and we can revisit a better design if/when the new WG-approved spec comes out with all those changes from the F2F.

**Comment by jmesserly** [comment]
sounds good to me

**Comment by azakus** [comment]
I made a minimal backport of the 0.7 ShadowDOM changes on top of 0.6.3

`git diff 0.6.3..master -- src/ShadowDOM | patch -Np1`

This seems sufficient to stop throwing on random `Window.prototype` accessors, probably due to #318 and a few other fixes (#318 was not sufficient by itself).

**Comment by TimvdLippe** [comment]
Thanks for the issue. It has been a great while since this issue was opened. Sadly, this issue misses a live reproduction with the most recent version of the webcomponents polyfill for us to be able to debug and potentially fix. Versions 1 of the polyfills fix a large amount of issues, probably including this one as well.

Please see [the CONTRIBUTING.md](https://github.com/webcomponents/webcomponentsjs/blob/master/CONTRIBUTING.md) for guidelines on filing issues.  You can start from [this jsBin template](http://jsbin.com/birajez/edit?html,console,output) as a starting point and link it in the issue template under "Live Demo".

Feel free to file a new issue following these instructions if you can still reproduce this issue with versions 1 of the polyfills.  Thank you!
